XMonad/Hooks/DynamicLog.hs

```diff
@@ -585,6 +589,103 @@ pprWindowSetXinerama ws = "[" ++ unwords onscreen ++ "] " ++ unwords offscreen
         offscreen = map S.tag . filter (isJust . S.stack)
                         . sortOn S.tag $ S.hidden ws
 
+-- | Custom datatype for custom icons based on the state of the 'Workspace'
+-- For example,
+--
+-- > Icon "<bold>discord</bold>" "<bold>discord</bold" "discord" ""
+--
+-- Then you can add it to your IconSet.
+--
+-- > icons :: IconSet
+-- > icons = mconcat
+-- >    [ className =? "discord" --> pure [Icon "<bold>discord</bold>" "<bold>discord</bold" "discord" ""]
+-- >    ]
+data Icon = Icon
+    { iconCurrent :: !String -- ^ If the 'Workspace' is the current workspace
+    , iconVisible :: !String -- ^ If the 'Workspace' is visible (Xinerama only)
+    , iconHidden :: !String -- ^ If the 'Workspace' isnt visible but still has windows
+    , iconHiddenNoWindows :: !String -- ^ If the 'Workspace' isnt visible and has no windows
+    }
+
+-- | The set of Icons to use
+type IconSet = Query [Icon]
+
+baseIconSet :: String -> Icon
+baseIconSet x =
+    Icon { iconCurrent = x
+         , iconVisible = x
+         , iconHidden = x
+         , iconHiddenNoWindows = x
+         }
+
+-- | Create an 'IconSet' from a 'String'
+appIcon :: String -> IconSet
+appIcon x = pure [baseIconSet x]
+
+-- | Dynamically changes 'WorkspaceId' based on the 'Window's inside the Workspace.
+-- 'IconSet's describe which icons are shown depending on which windows fit a query
+-- To create an 'IconSet' make a 'Query' that returns a '[Icon]'.
+-- 'appIcon' can be used to simplify this process
+-- For example,
+--
+-- > icons :: IconSet
+-- > icons = composeAll
+-- >           [ className =? "discord" --> appIcon "\xfb6e"
+-- >           , className =? "Discord" --> appIcon "\xf268"
+-- >           , className =? "Firefox" --> appIcon "\63288"
+-- >           , className =? "Spotify" <||>  className =? "spotify" --> appIcon "阮"
+-- >           ]
+--
+-- then you can add the hook to your config
+--
+-- > xmonad $ def
+-- >      {
+-- >      logHook = dynamicLogIconsWithPP icons xmobarPP <> myManageHook
+-- >      }
+--
+dynamicLogIconsWithPP :: IconSet -- ^ The 'IconSet' to use
+                      -> PP -- ^ The 'PP' to alter
+                      -> X () -- ^ The resulting 'X' action
+dynamicLogIconsWithPP iconset pp = dynamicLogWithPP =<< dynamicLogIconsConvert (def{ iconConfigIcons = iconset, iconConfigPP = pp})
```

Comment:
```suggestion
dynamicLogIconsWithPP iconset pp = dynamicLogWithPP =<< dynamicLogIconsConvert (def { iconConfigIcons = iconset, iconConfigPP = pp })
```

Reply:
I actually suggested this be changed from `def { ... }` to `def{ ... }` :>

I think that, due to record syntax binding very strongly, this makes it much easier to parse that this is a single thing (that way one may even get away with omitting those parentheses).